geometry/parse-ply.cpp

```diff
@@ -0,0 +1,316 @@
+        LOGTHROW(err4, std::runtime_error)
+            << "Unexpected EOF reached while parsing PLY header";
+    }
+}
+
+/// Helper retrieving the next line of file, throws if EOF reached
+inline void nextLine(std::ifstream& f, std::string& line)
+{
+    std::getline(f, line);
+    ba::trim(line);
+    throwOnEof(f);
+}
+
+/// Splits property line to name and type
+bool parseProperty(const std::string& line, std::vector<PlyProperty>& props)
+{
+    if (!ba::starts_with(line, "property")) { return false; }
+
+    std::size_t a = line.find_first_of(' ');
+    std::size_t b = line.find_last_of(' ');
+    if ((a == std::string::npos) || (a == b) || (b == line.size() - 1))
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Error parsing property line: " << line;
+    }
+
+    props.emplace_back(line.substr(a + 1, b - a - 1),          // prop type
+                       line.substr(b + 1, line.size() - b - 1) // prop name
+
+    );
+    return true;
+}
+
+/// Checks if a line contains a comment and optionally parses it
+bool parseComment(const std::string& line, std::vector<std::string>& comments)
+{
+    if (!ba::starts_with(line, "comment")) { return false; }
+
+    comments.push_back(line.substr(7));
+    ba::trim(comments.back()); // trim leading spaces
+    return true;
+}
+
+/// Parses lines of one element (vertex, face, ...) in header
+/// NB: After return, line contains a beginning of the next property, comment,
+/// or "end_header"
+bool parseElement(std::ifstream& f,
+                  std::string& line,
+                  std::vector<PlyElement>& elements,
+                  std::vector<std::string>& comments)
+{
+    if (!ba::starts_with(line, "element")) { return false; }
+
+    std::string keyword, name;
+    std::size_t count;
+    if (!(std::istringstream(line) >> keyword >> name >> count))
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Error parsing element defined by: " << line;
+    }
+
+    elements.emplace_back(name, count);
+    PlyElement& el = elements.back();
+    // parse the element properties
+    do
+    {
+        nextLine(f, line);
+    }
+    while (parseProperty(line, el.props) || parseComment(line, comments));
+
+    if (el.props.empty())
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Encountered element with no properties: " << name;
+    }
+
+    return true;
+}
+
+/// Check "format" line in PLY file, throws on unexpected inputs
+void checkFormatLine(std::ifstream& f)
+{
+    std::string line;
+    nextLine(f, line); // read whole line
+
+    std::string keyword, format, version;
+    if (!(std::istringstream(line) >> keyword >> format >> version))
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Error parsing format line: " << line;
+    }
+    if (keyword != "format")
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Expected the second line to contain \"format\" specification";
+    }
+    if (format != "ascii")
```

Comment:
If we are checking format/keyword, maybe also check the version, because in the future versions the format of the PLY file might change.